### ISSUE TITLE
TvDbSharper.TvDbServerException: Your JWT token is missing or expired; Not authorized

### ISSUE THREAD

**Issue by SenorSmartyPants** [jellyfin/jellyfin#1356]
**Describe the bug**
No new TVDB data. Lots of JWT token expiration messages in the log.

**Logs**
 

```
 [2019-05-05 19:31:19.418 -05:00] [ERR] Error in DummySeasonProvider
TvDbSharper.TvDbServerException: Your JWT token is missing or expired; Not authorized
   at TvDbSharper.Infrastructure.Parser.Parse[T](ApiResponse response, IReadOnlyDictionary`2 errorMap)
   at TvDbSharper.Clients.SeriesClient.GetEpisodesAsync(Int32 seriesId, Int32 page, EpisodeQuery query, CancellationToken cancellationToken)
   at MediaBrowser.Providers.TV.TheTVDB.TvDbClientManager.TryGetValue[T](String key, String language, Func`1 resultFactory)
   at MediaBrowser.Providers.TV.TheTVDB.TvDbClientManager.GetAllEpisodesAsync(Int32 tvdbId, String language, CancellationToken cancellationToken)
   at MediaBrowser.Providers.TV.MissingEpisodeProvider.Run(Series series, Boolean addNewItems, CancellationToken cancellationToken)
   at MediaBrowser.Providers.TV.SeriesMetadataService.AfterMetadataRefresh(Series item, MetadataRefreshOptions refreshOptions, CancellationToken cancellationToken)
[2019-05-05 19:31:19.557 -05:00] [INF] Creating Season "Season Unknown" entry for "The Simpsons"
[2019-05-05 19:31:20.057 -05:00] [ERR] Error in DummySeasonProvider
TvDbSharper.TvDbServerException: Your JWT token is missing or expired; Not authorized
   at TvDbSharper.Infrastructure.Parser.Parse[T](ApiResponse response, IReadOnlyDictionary`2 errorMap)
   at TvDbSharper.Clients.SeriesClient.GetEpisodesAsync(Int32 seriesId, Int32 page, EpisodeQuery query, CancellationToken cancellationToken)
   at MediaBrowser.Providers.TV.TheTVDB.TvDbClientManager.TryGetValue[T](String key, String language, Func`1 resultFactory)
   at MediaBrowser.Providers.TV.TheTVDB.TvDbClientManager.GetAllEpisodesAsync(Int32 tvdbId, String language, CancellationToken cancellationToken)
   at MediaBrowser.Providers.TV.MissingEpisodeProvider.Run(Series series, Boolean addNewItems, CancellationToken cancellationToken)
   at MediaBrowser.Providers.TV.SeriesMetadataService.AfterMetadataRefresh(Series item, MetadataRefreshOptions refreshOptions, CancellationToken cancellationToken)
[2019-05-05 19:31:24.284 -05:00] [ERR] Failed to retrieve episode with id null
TvDbSharper.TvDbServerException: Your JWT token is missing or expired; Not authorized
   at TvDbSharper.Infrastructure.Parser.Parse[T](ApiResponse response, IReadOnlyDictionary`2 errorMap)
   at TvDbSharper.Clients.SeriesClient.GetEpisodesAsync(Int32 seriesId, Int32 page, EpisodeQuery query, CancellationToken cancellationToken)
   at MediaBrowser.Providers.TV.TheTVDB.TvDbClientManager.TryGetValue[T](String key, String language, Func`1 resultFactory)
   at MediaBrowser.Providers.TV.TheTVDB.TvDbClientManager.GetEpisodeTvdbId(Int32 seriesTvdbId, EpisodeQuery episodeQuery, String language, CancellationToken cancellationToken)
   at MediaBrowser.Providers.TV.TheTVDB.TvdbEpisodeProvider.GetMetadata(EpisodeInfo searchInfo, CancellationToken cancellationToken)
```

**System (please complete the following information):**
 - OS: Docker
 - Jellyfin Version: 10.3.2

**Comment by anthonylavado** [comment]
Strange - I'm not getting this right now. I know the TVDB has been changing some infrastructure lately.

It looks like you have the "Display missing episodes within seasons" feature on. Is that correct?

**Comment by SenorSmartyPants** [comment]
"Display missing episodes within seasons" feature is off.

**Comment by anthonylavado** [comment]
Hrm... There's not many reasons why the DummySeasonProvider would be called otherwise. Keep an eye on it and let me know if it comes up again. I've been hammering The TVDB a lot lately with tests, turning different things off and on, and I haven't had this one yet.

**Comment by anthonylavado** [comment]
Okay, one other person with the error:

https://github.com/jellyfin/jellyfin/issues/1282#issuecomment-490209275

@SenorSmartyPants @kusochi Can you both please provide the settings you have in the TV Shows library?

**Comment by kusochi** [comment]
@anthonylavado My settings:

### Library Settings
Preferred download language: none selected
Country: none selected
Special season display name: Specials
Prefer embedded titles over filenames: disabled
Enable real time monitoring: enabled
Series metadata downloaders: TheTVDB and TheMoviesDb
Season metadata downloaders: none
Episode metadata downloaders: TheTVDB
Automatically refresh metadata from the internet: Never
Metadata savers: none
Series Image Fetchers: TheTVDB and FanArt
Season Image Fetchers: TheTVDB
Episode Image Fetchers: TheTVDB and Screen Grabber
Save artwork into media folders: disabled
Download images in advance: disabled
Automatically merge series that are spread across multiple folders: enabled
Display missing episodes within seasons: disabled
### Chapter Images
Enable chapter image extraction: disabled
Extract chapter images during the library scan: disabled
### Subtitle Downloads
Download languages: none
Subtitle downloaders: Open Subtitles
Only download subtitles that are a perfect match for my video files: enabled
Skip if the default audio track matches the download language: disabled
Skip if the video already contains embedded subtitles: disabled
Save subtitles into media folders: enabled

My installation was originally migrated from Emby, if that is relevant.

**Comment by SenorSmartyPants** [comment]
My library settings are the same as the above except where I'm noting below.

### Library Settings
Preferred download language: English
Country: United States


I don't have a subtitle plugin installed.

Should there be a TVDB optioin in Season metadata downloaders?

**Comment by dkamanae** [comment]
I'm having the same issue with the same library settings. I also migrated from Emby originally, just in case that's a pattern.

**Comment by anthonylavado** [comment]
@SenorSmartyPants No TVDB option in Season Metadata Downloaders. Are you also migrated from Emby?

**Comment by HotMykeul** [comment]
Same problem for me.

`[2019-05-10 03:29:04.725 +00:00] [ERR] Failed to retrieve series with id null
TvDbSharper.TvDbServerException: Your JWT token is missing or expired; Not authorized
   at TvDbSharper.Infrastructure.Parser.Parse[T](ApiResponse response, IReadOnlyDictionary`2 errorMap)
   at TvDbSharper.Clients.SeriesClient.GetAsync(Int32 seriesId, CancellationToken cancellationToken)
   at MediaBrowser.Providers.TV.TheTVDB.TvDbClientManager.TryGetValue[T](String key, String language, Func`1 resultFactory)
   at MediaBrowser.Providers.TV.TheTVDB.TvdbSeriesProvider.FetchSeriesData(MetadataResult`1 result, String metadataLanguage, Dictionary`2 seriesProviderIds, Canc$`

**Comment by anthonylavado** [comment]
A question going forward to all concerned:
Have you migrated settings/libraries from Emby at all? Or did you start Jellyfin "from scratch"?

**Comment by HotMykeul** [comment]
Migrated from emby for me.

Le ven. 10 mai 2019 à 06:16, Anthony Lavado <notifications@github.com> a
écrit :

**Comment by SenorSmartyPants** [comment]
I ran emby first, but did not migrate as I ran both in parallel. I ran them in separate dockers on the same machine.

**Comment by kusochi** [comment]
I wiped my installation and got a installation "from scratch" with fresh configuration and libraries and the issue is not present any more on the same machine with same settings.

**Comment by SenorSmartyPants** [comment]
That's what I assumed would happen but I'd rather not do that. This is definitely a problem with not renewing the authorization with TVDB as I see it.

Is there a less extreme way to reset the TVDB config than doing a fresh install? Maybe just a couple config files to wipe?

**Comment by HotMykeul** [comment]
> That's what I assumed would happen but I'd rather not do that. This is definitely a problem with not renewing the authorization with TVDB as I see it.
> 
> Is there a less extreme way to reset the TVDB config than doing a fresh install? Maybe just a couple config files to wipe?

Completely agree, I do not want to do a fresh install, I will loose all viewed movies and tv, and manually identified movies.

**Comment by cvium** [comment]
Restart jellyfin

**Comment by SenorSmartyPants** [comment]
I restarted JF and am doing a rescan for missing metadata. TVDB is responding and data is coming in.

**Comment by HotMykeul** [comment]
I restarted it (docker restart jellyfin), it do not work for me.

**Comment by dkamanae** [comment]
Restarting Jellyfin also did not work for me.

**Comment by anthonylavado** [comment]
@HotMykeul Could you try the "shutdown" option in the Dashboard, then try restarting the container? I'm thinking that it's still keeping whatever data around, and not actually refreshing it properly (though I don't use Docker, so I can't really comment)

@dkamanae Could you also try the "shutdown" button? How did you install Jellyfin?

**Comment by dkamanae** [comment]
I install via the repo for Debian.
I used the shutdown option, and it wasn't immediately working. But, I checked the log and the error was no longer showing up, so I looked in the library settings and thetvdb had been unselected as a source for episode data. I rechecked it, and now it's working as expected! Thanks!

**Comment by DrPandemic** [comment]
Is there a way to run the full shutdown process when a docker container gets killed?

**Comment by SenorSmartyPants** [comment]
This started happening again last night. I shutdown JF from the dashboard and restarted the docker container. Pulling data from TVDB again. But it's annoying to have to watch for this issue.

**Comment by anthonylavado** [comment]
@senorsmartypants
In the Scheduled Tasks area of the Dashboard, do you have “Scan media library” active? Does it run every 12 hours?

**Comment by anthonylavado** [comment]
Some discussion on Riot/Matrix says that we need to add some error handling around TheTVDB.

My personal guess without knowing anymore details is that we’re not getting a response (either because the API is busy/down), and then the token becomes invalid as a result.

**Comment by SenorSmartyPants** [comment]
Scan media library is active. It is set to run every 12 hours.

I don't recall if I ever touched that setting or not.

**Comment by anthonylavado** [comment]
@SenorSmartyPants Looks like @cvium got it in #1426. There was a comparison set the wrong way, so it would always refresh the key when it was too late.

**Comment by SenorSmartyPants** [comment]
Great. Any ETA on the next release?

**Comment by anthonylavado** [comment]
Not sure. I’ve just merged it in to the release branch, and I think we’re almost ready.

**Comment by jomp16** [comment]
It still isn't working...

Jellyfin docker image hash: b0a42f5a8ef31e5bb95f774e2a9c758b33ff838eb7b701af62581adc9787c95f
Version: 10.3.5

```
[22:21:42] [ERR] Error in DummySeasonProvider
TvDbSharper.TvDbServerException: Your JWT token is missing or expired; Not authorized
   at TvDbSharper.Infrastructure.Parser.Parse[T](ApiResponse response, IReadOnlyDictionary`2 errorMap)
   at TvDbSharper.Clients.SeriesClient.GetEpisodesAsync(Int32 seriesId, Int32 page, EpisodeQuery query, CancellationToken cancellationToken)
   at MediaBrowser.Providers.TV.TheTVDB.TvDbClientManager.TryGetValue[T](String key, String language, Func`1 resultFactory)
   at MediaBrowser.Providers.TV.TheTVDB.TvDbClientManager.GetAllEpisodesAsync(Int32 tvdbId, String language, CancellationToken cancellationToken)
   at MediaBrowser.Providers.TV.MissingEpisodeProvider.Run(Series series, Boolean addNewItems, CancellationToken cancellationToken)
   at MediaBrowser.Providers.TV.SeriesMetadataService.AfterMetadataRefresh(Series item, MetadataRefreshOptions refreshOptions, CancellationToken cancellationToken)
[22:21:42] [INF] Scan media library Completed after 1 minute(s) and 9 seconds
```

**Comment by SenorSmartyPants** [comment]
I'm still getting this error as well on 10.3.5. doing a docker restart on the JF image made it start working again.

**Comment by dkamanae** [comment]
Unfortunately, I have to report I'm still getting this on 10.3.5, doing a shutdown from the dashboard and then restarting seems to fix it temporarily, but eventually it starts happening again.

**Comment by cvium** [comment]
This will be fixed in 10.3.6.

**Comment by enomam** [comment]
I've just upgraded to 10.3.6 (docker image), and am now getting these errors. Related?

```
[2019-07-07 06:18:54.793 +00:00] [ERR] Error in DummySeasonProvider
System.NullReferenceException: Object reference not set to an instance of an object.
   at TvDbSharper.Infrastructure.ApiClient.GetResponseAsync(WebRequest httpRequest, CancellationToken cancellationToken)
   at TvDbSharper.Infrastructure.ApiClient.SendRequestAsync(ApiRequest request, CancellationToken cancellationToken)
   at TvDbSharper.Clients.AuthenticationClient.AuthenticateAsync(AuthenticationData authenticationData, CancellationToken cancellationToken)
   at MediaBrowser.Providers.TV.TheTVDB.TvDbClientManager.get_TvDbClient()
   at MediaBrowser.Providers.TV.TheTVDB.TvDbClientManager.<>c__DisplayClass18_0.<GetEpisodesPageAsync>b__0()
   at MediaBrowser.Providers.TV.TheTVDB.TvDbClientManager.TryGetValue[T](String key, String language, Func`1 resultFactory)
   at MediaBrowser.Providers.TV.TheTVDB.TvDbClientManager.GetAllEpisodesAsync(Int32 tvdbId, String language, CancellationToken cancellationToken)
   at MediaBrowser.Providers.TV.MissingEpisodeProvider.Run(Series series, Boolean addNewItems, CancellationToken cancellationToken)
   at MediaBrowser.Providers.TV.SeriesMetadataService.AfterMetadataRefresh(Series item, MetadataRefreshOptions refreshOptions, CancellationToken cancellationToken)
```

**Comment by anthonylavado** [comment]
We have some HTTP stuff that needs to be fixed up, we'll try to release soon. Not 100% sure if this is related.

**Comment by tyhawk** [comment]
I've got the same error as @enomam, but on 10.3.7 (also running docker).
```
[2019-09-19 10:53:49.716 +02:00] [ERR] Error in DummySeasonProvider
System.NullReferenceException: Object reference not set to an instance of an object.
   at TvDbSharper.Infrastructure.ApiClient.GetResponseAsync(WebRequest httpRequest, CancellationToken cancellationToken)
   at TvDbSharper.Infrastructure.ApiClient.SendRequestAsync(ApiRequest request, CancellationToken cancellationToken)
   at TvDbSharper.Clients.AuthenticationClient.AuthenticateAsync(AuthenticationData authenticationData, CancellationToken cancellationToken)
   at MediaBrowser.Providers.TV.TheTVDB.TvDbClientManager.get_TvDbClient()
   at MediaBrowser.Providers.TV.TheTVDB.TvDbClientManager.<>c__DisplayClass18_0.<GetEpisodesPageAsync>b__0()
   at MediaBrowser.Providers.TV.TheTVDB.TvDbClientManager.TryGetValue[T](String key, String language, Func`1 resultFactory)
   at MediaBrowser.Providers.TV.TheTVDB.TvDbClientManager.GetAllEpisodesAsync(Int32 tvdbId, String language, CancellationToken cancellationToken)
   at MediaBrowser.Providers.TV.MissingEpisodeProvider.Run(Series series, Boolean addNewItems, CancellationToken cancellationToken)
   at MediaBrowser.Providers.TV.SeriesMetadataService.AfterMetadataRefresh(Series item, MetadataRefreshOptions refreshOptions, CancellationToken cancellationToken)
```
Scanning my libraries since today is extremely slow. Haven't touched any settings. Not sure what's going on.

**Comment by anthonylavado** [comment]
Ah okay, thanks for letting us know. This should be a new issue if it isn't already. It's not related to the JWT token.
